### PR TITLE
Migrate automatic python resource list to Homebrew/core

### PR DESCRIPTION
**Pull request by Rylan12**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

Supersedes #8770 and #9143

This PR adds three flags to `update-python-resources`:

1. `--package-name` allows for an alternative PyPI package to be specified for finding dependencies. This is helpful for situations where the formula name is `foo` but the resources required come from `foo[abc]` (or similar situations).
1. `--extra-packages` allows for a list of additional PyPI packages to be specified. This is helpful for instances where `pipgrip` doesn't pick up all of the needed dependencies. This can include packages like `foo[abc]` and `foo==1.2.3` for specific versions.
1. `--exclude-packages` allows for certain packages to be excluded from the generated resource blocks. This is useful for instances where `pipgrip` finds dependencies that we don't want to be included as resource blocks.

I've also changed the logic for the auto-update blocklist to migrate to Homebrew/core and add more functionality to utilize the new flags.

Corresponding Homebrew/core PR: https://github.com/Homebrew/homebrew-core/pull/65096

One question: should/can we rename the `audit_exceptions/` directory (and methods that are similarly named) to a more accurate name (maybe `formula_lists` or similar). This is an example of a list being added that doesn't relate to audits. I'm not sure whether there will be an issue with compatibility if I rename it so I've left it as is for now.

### CODE REVIEW

File: Library/Homebrew/utils/pypi.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/pypi.rb:56]: Pedantry: this isn't an `audit_exception` in any meaningful form. I think it's probably still worth migrating to Homebrew/core but with a different name/location.

File: Library/Homebrew/utils/pypi.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/pypi.rb:82]:
```suggestion
```

**Review by MikeMcQuaid — COMMENTED** [review]
Code seems good to me. I'd like to see this use a different file/directory as this isn't technically related to audits.

File: Library/Homebrew/utils/pypi.rb
Comment (Rylan12) [reply Library/Homebrew/utils/pypi.rb:56]: Agreed. I suggested `formula_lists` above but am 100% open to other names.

Also, just want to make sure it's okay to rename that in terms of compatibility and whatnot

File: Library/Homebrew/utils/pypi.rb
Comment (Rylan12) [reply Library/Homebrew/utils/pypi.rb:82]: Ugh... I even wrote myself a note to remove these 🤦‍♂️

File: Library/Homebrew/utils/pypi.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/utils/pypi.rb:56]: Maybe another dedicated directory (or file in the root) just for the Python resource stuff?

Comment (Rylan12) [reply Library/Homebrew/utils/pypi.rb:56]: Let's move the discussion to https://github.com/Homebrew/brew/pull/9178 (which I'm repurposing for some refactoring of this logic)

**Comment by Rylan12** [comment]
Putting this on hold until the migration in https://github.com/Homebrew/brew/pull/9178 is finished.

File: Library/Homebrew/utils/pypi.rb
Comment (dtrodrigues) [inline Library/Homebrew/utils/pypi.rb:54]: Since this changed from `version = nil` to `version: nil`, `bump-formula-pr.rb` line 339 needs to be changed to
```
resources_checked = PyPI.update_python_resources! formula, version: new_formula_version,
```
to support for instance, `brew bump-formula-pr --force -w ipython --version 7.19.0`

Comment (Rylan12) [reply Library/Homebrew/utils/pypi.rb:54]: Ah, good catch, thanks!

(Changed in the `update-python-resources` command but not in `bump-formula-pr`)

**Review by dtrodrigues — APPROVED** [review]

**Comment by Rylan12** [comment]
Okay, now that #9178 is done, this PR is up. The main point of discussion: where should the PyPI resource auto-update list live in Homebrew/core? The options are as follows:

1. This list (and all other non-audit and non-style lists for e.g. `livecheck`) will live in a single `formula_lists/` directory
    - Pros: Reduces clutter in Homebrew/core. Adding more lists in the future won't require modifying any `tap` brew code.
    - Cons: Unrelated lists end up grouped together (especially if more lists are added in the future)
1. Have each type of list belong to its own directory. For example, this list could go in `pypi_lists/` while the `livecheck` list could go in `livecheck_lists/`
    - Pros: Better organization because each module/group has it's own directory
    - Cons: Adds more clutter to the tap's root directory. At the moment, though, this is only adding one additional directory compared to option 1. In the future, though, if more lists are added, this would mean adding more directories. Additionally, adding new list categories in the future will require some extra modification of `tap` brew code.
1. Have each non-audit and non-style list file live in the root directory (not in a folder)
    - Pros: None?
    - Cons: Increase clutter
1. Any of the above options but instead of having the list directories in the tap's root directory, nesting all directories/files in an e.g. `tap_lists/` directory.
    - Pros: Further reduces clutter because all lists are nested under a single directory
    - Cons: May be an unnecessary improvement at the moment. Also, while it decreases clutter, it makes it more difficult to navigate the tap file structure.

---

My opinions:

No for number 3
No for number 4 (I like the organization, but I don't think it's a big issue so why bother)
I'm a little torn between options 1 and 2. I think in the current state (i.e. there are only two non-audit and non-style lists), option 2 is totally fine. If we add more in the future, though, option 1 may make more sense. Not sure. I'm leaning toward option 2 at the moment. We can always change in the future.

**Comment by MikeMcQuaid** [comment]
> Have each type of list belong to its own directory. For example, this list could go in `pypi_lists/` while the `livecheck` list could go in `livecheck_lists/`
> Have each non-audit and non-style list file live in the root directory (not in a folder)

I am happy with either of these, leaning towards the latter unless we expect to have multiple "constants" (i.e. whatever we replace them with) for each.

**Comment by Rylan12** [comment]
> I am happy with either of these, leaning towards the latter unless we expect to have multiple "constants" (i.e. whatever we replace them with) for each.

Interesting. Yes, there's one list for each of those at the moment. Originally I was definitely 👎 on this but not sure. I guess it's not really any different than the `tap_migrations` or `formula_renames` files, so I'd be open to it. (I've updated the pros/cons list as I was a little too harsh on this option)

**Comment by MikeMcQuaid** [comment]
> Originally I was definitely 👎 on this but not sure. I guess it's not really any different than the `tap_migrations` or `formula_renames` files, so I'd be open to it. (I've updated the pros/cons list as I was a little too harsh on this option)

Yeh, that's my thinking!

**Comment by Rylan12** [comment]
Okay, thought about this more. I'm 👍 on that solution. Will update this PR (and the homebrew/core PR)

**Comment by Rylan12** [comment]
As discovered in https://github.com/Homebrew/homebrew-core/pull/65361#discussion_r528227777, `update-python-resources` should treat `_` and `-` the same in package names (as described in [PEP 426](https://www.python.org/dev/peps/pep-0426/#name))

To accomplish this, I did some refactoring of the way we handle packages. I also added type signatures to the methods in `utils/pypi`. However, setting `# typed: true` caused `brew typecheck` to fail for what appears to be a lack of type definitions in the `Formula` class, so I left it as `# typed: false`.

**Comment by dtrodrigues** [comment]
I tested the latest locally and it's working as expected for both `youtube-dl` and `pre-commit`, which was also affected by the `-` vs `_` handling.

**Comment by Rylan12** [comment]
Alright, assuming there are no major concerns raised in the next few hours I'm going to go ahead and merge. Next up, I plan on moving the rest of the lists to homebrew/core. My current plan is to do it in three stages:

1. The `livecheck` list
1. The remaining audit exception lists
1. The style exception lists

Those can be discussed in more detail in their respective PRs.

**Comment by reitermarkus** [comment]
> go ahead and merge

You'll need to rebase this so the `review` workflow runs and passes.

**Comment by Rylan12** [comment]
Hmm, CI is now failing with some new (and unrelated) issues. Looks like some problems with the `openjdk@8` migration in https://github.com/Homebrew/homebrew-core/pull/65507 and with `brew audit` failing for casks with no description...

**Comment by Rylan12** [comment]
Rebased after https://github.com/Homebrew/brew/pull/9275 to get CI passing. Will merge once ✅

**Comment by vinceatbluelabs** [comment]
Hi @Rylan12!  Thanks for pushing this one through!  Upon testing it in my private tap, I encountered the problem below.  Does this give you enough of a hint to figure out what's going on, or would you like me to cook up reproducible steps and file a separate issue?

```sh
$ brew update-python-resources --package-name 'records-mover[cli,pandas,literally_every_single_database_binary,gsheets]' mover-cli.rb
time brew update-python-resources --package-name 'records-mover[cli,pandas,literally_every_single_database_binary,gsheets]' mover-cli.rb
Error: undefined method `pypi_formula_mappings' for nil:NilClass
Please report this issue:
  https://docs.brew.sh/Troubleshooting
/usr/local/Homebrew/Library/Homebrew/utils/pypi.rb:133:in `update_python_resources!'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/update-python-resources.rb:45:in `block in update_python_resources'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/update-python-resources.rb:44:in `each'
/usr/local/Homebrew/Library/Homebrew/dev-cmd/update-python-resources.rb:44:in `update_python_resources'
/usr/local/Homebrew/Library/Homebrew/brew.rb:119:in `<main>'
$ 
```

**Comment by Rylan12** [comment]
@vinceatbluelabs thanks for reporting! No need for another issue, I can reproduce this (it's pretty much the same issue that was reported in https://github.com/Homebrew/brew/pull/9249).

I could have sworn I checked to made sure to check that it wouldn't be an issue here, but clearly, not thoroughly enough... 🤦‍♂️  I've opened https://github.com/Homebrew/brew/pull/9283/files to fix it (that PR should be good to merge once CI passes which should be pretty quickly).

**Comment by Rylan12** [comment]
Okay, fixed. Let me know if you're still having issues

**Pull request merged**
